### ISSUE TITLE
tracking issue topics Smit

### ISSUE THREAD

**Issue by rgommers** [rgommers/scipy#36]
Hi @Smit-create, as we talked about on the weekend, let's keep a list of topics and ideas that would be interesting to work on.

- [x] https://github.com/rgommers/scipy/issues/22 (`misc`, `io`, maybe there's more later)
- [x] https://github.com/rgommers/scipy/issues/30
- https://github.com/rgommers/scipy/issues/33 (this may be a challenging one)
- [x] fixing build warnings in the SciPy code base that we temporarily silenced - bugs could be hiding here: https://github.com/rgommers/scipy/issues/42
- [x] https://github.com/rgommers/scipy/issues/32
- [ ] Following up on the `signal.sepfir2d` issues: https://github.com/rgommers/scipy/issues/43
- ~meson: timestamp caching in CI trick. The `build/` folder can be saved and reused for future PRs, but the timestamps must be manipulated in a smart way in order to get rebuilds for only the targets that were affected by a PR and commits since the last rebuild. I read a blog post about this that I cannot find right now. (no separate issue yet)~
- [x] It would be interesting to learn how BLAS/LAPACK library affects performance on different CPUs, as well as the relative performance of BLAS libraries. There are blog posts about this, in particular for OpenBLAS and MKL on Intel and AMD CPUs. We're getting more interested in `aarch64` too. And haven't seen much about `arm64` (Mac M1) yet.
- [x] Similar to above, a little easier: how does build time and benchmark performance depend on optimization level (`-O0` ... `-O3`, `-g`, etc.).
- [x] PROPACK support in Meson build files (finishing https://github.com/rgommers/scipy/pull/62).
- [x] gh-69
- [x] gh-72
- [x] update the `meson` branch in this fork to be up-to-date with SciPy `master` (see gh-93)
- [x] add a `dev.py` interface for the Meson build: current idea is to import functions that don't need changes from `runtests.py` and reimplement `build_project`.
- [x] ensure `stats.linalg` tests are installed and passing (currently ~90 failures on master)
- [x] write a uarray backend for a small part of `scipy.signal` and create a Colab notebook which works with CuPy and https://docs.rapids.ai/api/cusignal/nightly/api.html#module-cusignal.filtering.resample
- [x] gh-120
- [x] Add a new section to the [SciPy docs on Meson](http://scipy.github.io/devdocs/dev/contributor/meson.html) about development topics, covering building with different optimization levels (closes gh-58) and building with ASAN (closes gh-32). 
- [x] Related to the above topic: write a new section on how to use several different builds in parallel: `gcc`, `clang` regular builds and a gcc debug build on Linux.
- [ ] Make Meson work with gcov code coverage: https://github.com/scipy/scipy/issues/15896
- [ ] sdist and wheels support via `meson-python`

**Comment by rgommers** [comment]
For comparing BLAS libraries on the same machine, here's how to easily switch them out with conda: https://conda-forge.org/docs/maintainer/knowledge_base.html#switching-blas-implementation

**Comment by Smit-create** [comment]
It produces the following error:
```
$ conda install "libblas=*=*mkl"
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.

PackagesNotFoundError: The following packages are not available from current channels:

  - libblas[build=*mkl]

Current channels:

  - https://repo.anaconda.com/pkgs/main/linux-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/r/linux-64
  - https://repo.anaconda.com/pkgs/r/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.

```
From `$ conda list`, I get:

```
# Name                    Version                   Build  Channel

libblas                   3.9.0                9_openblas    conda-forge
libcblas                  3.9.0                9_openblas    conda-forge
libopenblas               0.3.15          pthreads_h8fe5266_1    conda-forge
mkl                       2021.3.0           h06a4308_520  
```
Should there be something else to change the build type?

**Comment by rgommers** [comment]
Ah you're installing from `defaults` (see "Current channels" listed), this feature is specific to `conda-forge`.

**Comment by rgommers** [comment]
The solve is slow when done in an existing `scipy-dev` environment based on the `environment.yml` in SciPy master, but it does work:

<details>

```
$ conda install "libblas=*=*mkl"
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: done
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: done

## Package Plan ##

  environment location: /home/rgommers/anaconda3/envs/scipy-dev

  added / updated specs:
    - libblas[build=*mkl]


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    binutils_linux-64-2.36     |       hf3e587d_1          23 KB  conda-forge
    gcc_impl_linux-64-9.4.0    |       h03d3576_8        45.7 MB  conda-forge
    gcc_linux-64-9.4.0         |       h391b98a_1          24 KB  conda-forge
    gfortran_impl_linux-64-9.4.0|       h0003116_8        10.0 MB  conda-forge
    gfortran_linux-64-9.4.0    |       hf0ab688_1          23 KB  conda-forge
    gxx_impl_linux-64-9.4.0    |       h03d3576_8        10.6 MB  conda-forge
    gxx_linux-64-9.4.0         |       h0316aca_1          23 KB  conda-forge
    libblas-3.9.0              |   11_linux64_mkl          12 KB  conda-forge
    libcblas-3.9.0             |   11_linux64_mkl          11 KB  conda-forge
    libgcc-devel_linux-64-9.4.0|       hd854feb_8         4.0 MB  conda-forge
    liblapack-3.9.0            |   11_linux64_mkl          11 KB  conda-forge
    libsanitizer-9.4.0         |       h79bfe98_8         6.9 MB  conda-forge
    libstdcxx-devel_linux-64-9.4.0|       hd854feb_8         9.9 MB  conda-forge
    openssl-1.1.1l             |       h7f98852_0         2.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:        89.4 MB

The following NEW packages will be INSTALLED:

  libsanitizer       conda-forge/linux-64::libsanitizer-9.4.0-h79bfe98_8
  llvm-openmp        conda-forge/linux-64::llvm-openmp-12.0.1-h4bd325d_1
  mkl                conda-forge/linux-64::mkl-2021.3.0-h726a3e6_557
  tbb                conda-forge/linux-64::tbb-2021.3.0-h4bd325d_0

The following packages will be UPDATED:

  gcc_impl_linux-64                       9.3.0-h70c0ae5_19 --> 9.4.0-h03d3576_8
  gcc_linux-64                            9.3.0-hf25ea35_32 --> 9.4.0-h391b98a_1
  gfortran_impl_lin~                      9.3.0-hc4a2995_19 --> 9.4.0-h0003116_8
  gfortran_linux-64                       9.3.0-hdc58fab_32 --> 9.4.0-hf0ab688_1
  gxx_impl_linux-64                       9.3.0-hd87eabc_19 --> 9.4.0-h03d3576_8
  gxx_linux-64                            9.3.0-h3fbe746_32 --> 9.4.0-h0316aca_1
  libblas                                  3.9.0-9_openblas --> 3.9.0-11_linux64_mkl
  libcblas                                 3.9.0-9_openblas --> 3.9.0-11_linux64_mkl
  libgcc-devel_linu~                      9.3.0-h7864c58_19 --> 9.4.0-hd854feb_8
  libgcc-ng                               9.3.0-h2828fa1_19 --> 11.1.0-hc902ee8_8
  libgfortran-ng                          9.3.0-hff62375_19 --> 11.1.0-h69a702a_8
  libgfortran5                            9.3.0-hff62375_19 --> 11.1.0-h6c583b3_8
  libgomp                                 9.3.0-h2828fa1_19 --> 11.1.0-hc902ee8_8
  liblapack                                3.9.0-9_openblas --> 3.9.0-11_linux64_mkl
  libstdcxx-devel_l~                      9.3.0-hb016644_19 --> 9.4.0-hd854feb_8
  libstdcxx-ng                            9.3.0-h6de172a_19 --> 11.1.0-h56837e0_8
  openssl                                 1.1.1k-h7f98852_0 --> 1.1.1l-h7f98852_0

The following packages will be DOWNGRADED:

  _openmp_mutex                                   4.5-1_gnu --> 4.5-1_llvm
  binutils_linux-64                        2.36-hf3e587d_32 --> 2.36-hf3e587d_1
```

</details>

**Comment by Smit-create** [comment]
Hi,
These are the results on my machine:
```
Library                                           Real time                       user time              sys time
----------------------------------------------------------------------------------------------------------------------
openblas                                         10m 9.007s                      19m 14.320s            7m 5.249s
mkl                                              7m 46.486s                      10m 54.084s            0m 28.639s
blis                                             12m 2.356s                      17m 40.777s            7m 2.029s
netlib                                           12m 45.545s                     20m 10.659s            7m 47.239s
```

My CPU details are:
```
"arch": "x86_64",
"cpu": "Intel(R) Core(TM) i3-4005U CPU @ 1.70GHz",
"num_cpu": "4",
"os": "Linux 5.4.0-81-generic",
"ram": "3950024",
```

The python script which I used for obtaining the results can be found at: https://gist.github.com/Smit-create/99f7fb2b8781053c39c3f25215e26cbf

**Comment by rgommers** [comment]
The "deprecate private API in SciPy" issue is coming to an end, we discussed to after that update the `meson` branch in this fork to be up-to-date with SciPy master.

**Comment by rgommers** [comment]
Old issue, closing.
